### ISSUE TITLE
"Set email" feature workflow doesn't result in changed email

### ISSUE THREAD

**Issue by gaieges** [LosFuzzys/MediaWiki-OAuth2-Github#9]
This is strange problem that I just ran into with (I believe) the plugin.  I'm trying to go through the change email workflow, which results in this sequence:

- On preferences page, click change email
- Redirected to `Special:ChangeEmail`, which returns a location header that forwards to `Special:UserLogin`
- Click login with github, go to github, go back to callback properly
- Cookies set properly, Get redirected to `Special:ChangeEmail` again
- `Special:ChangeEmail` gets sent to `Special:UserLogin` again

Any thoughts as to what's going on here?

MediaWiki `1.28.2 (438c3d6)19:03, 30 April 2017`
PHP `7.0.20-2 (apache2handler)`
MySQL `5.7.11-log`
OAuth2Github `0.0.2 (ebb48c2)15:05, 7 July 2017`

**Comment by stefan2904** [comment]
No spontaneous ideas, sorry.

So this does happen when you login via Github and then want to change the email address of that (Github-provided) user, not a local Mediawiki user?

**Comment by PeterTheOne** [comment]
Maybe it was never considered that GitHub users change their Email.. I'm not sure if this is a good idea. :D

**Comment by stefan2904** [comment]
@PeterTheOne we are not using that feature, so I am not sure. But I can think of cases where you would want to set your mediawiki password (for notifications, for example).

One theory I have is that the way mediawiki crates accounts for people logging in via this plugin does not allow changing the local email address. But I currently dont have time to verify that, sorry ...

**Comment by gaieges** [comment]
Yes @stefan2904 - attempting to simply change the github-oauth authenticated users email, in mediawiki. 

What we're trying to do is simply ensure that an email address exists in mediawiki so that each user can get emails on changes to pages, for example.  I'd be happy to not have to make users go through that process to click the button to 'change their email', and rather just have mediawiki have that info on first login or something, but I'm not sure what the workflow should be.

The change you made the other day IIRC just adds a line that add's the user if its not there - does that function touch email at all?  If not, I may be able to take a stab at adding another step after that to pull the email in locally as well.  I have no experience with wordpress dev so I'll want your guidance / review.

**Comment by stefan2904** [comment]
One could ask github for the email address (the same way we query the github API for group memberships), but that is not implemented right now.

**Comment by stefan2904** [comment]
There is a step in the registration process that sets the email address to the one provided by github:
https://github.com/LosFuzzys/MediaWiki-OAuth2-Github/blob/master/SpecialOAuth2Github.php#L221

Apparently thats not working in your installation?

**Comment by gaieges** [comment]
I guess not?  Or maybe its just not getting it from github on initial attempt.  I can play with this today and see what I can do.  I'll update the ticket and see if I can get it resolved.

**Comment by gaieges** [comment]
Ahh I see what's going on .. supposedly the object that comes back from github doesn't contain the 'email' field *if* your email is set to private.  My config (per the example) is just using `https://api.github.com/user` as the `info_endpoint`.  So I presume what's going on in that case is that the OAuth2 Client is just using that public endpoint to collect that info, which is generally anonymous (idk if it is in this case).

Do we need to make an authenticated call to get the private email?

**Comment by stefan2904** [comment]
no, we already request `user:email`, but apparently you need to make another api call if you want the non-private email (See https://developer.github.com/v3/users/emails/ )
